mantidproject/mantid
# Fix various deadlocks on ProjectRecovery shutdown

**Description of work.**
Fixes various deadlocks in ProjectRecovery:

- Scopes a critical section correctly, so we don't call into the GUI thread which is also waiting for the critical section
- Changes thread joining to detaching so we don't block the GUI (this also means we can get a `CancelException`
- Fixes a problem with workspaces being removed from the ADS before we are ready

**To test:**
- Open a workspace group with a large amount of history (or duplicate a workspace many times), I have an example if required
- Set the logger to debug
- Wait for project saving to start
- Before it finishes close Mantid normally
- Ensure it does not deadlock

----------------------
- Run the script attached to this issue #22714
- Wait a period of time. There should be no warnings about a workspace not being found

Fixes #22714 and Fixes #22736 

Does this update require release notes?

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
